File: src/nishikihebi/logs.py
import json
import logging
from datetime import UTC, datetime
from pathlib import Path
from typing import TextIO

from nishikihebi.console import BOLD, CYAN, DIM, RED, YELLOW, color_enabled, style

_LEVEL_CODES: dict[str, tuple[str, ...]] = {
    "DEBUG": (DIM,),
    "INFO": (CYAN,),
    "WARNING": (YELLOW,),
    "ERROR": (RED,),
    "CRITICAL": (BOLD, RED),
}


class ColorFormatter(logging.Formatter):
    def __init__(self, stream: TextIO) -> None:
        super().__init__()
        self.stream = stream

    def format(self, record: logging.LogRecord) -> str:
        levelname = f"{record.levelname:<7}"
        codes = _LEVEL_CODES.get(record.levelname, ())
        if codes and color_enabled(self.stream):
            levelname = style(levelname, *codes, stream=self.stream)
        return f"{levelname} {record.getMessage()}"


class JsonLinesFormatter(logging.Formatter):
    def format(self, record: logging.LogRecord) -> str:
        payload = {
            "time": datetime.fromtimestamp(record.created, tz=UTC).isoformat(),
            "level": record.levelname,
            "logger": record.name,
            "message": record.getMessage(),
            **getattr(record, "context", {}),
        }
        return json.dumps(payload)


def configure_logging(
    directory: Path = Path("log"), timestamp: datetime | None = None
) -> Path:
    timestamp = timestamp or datetime.now(UTC)
    directory.mkdir(parents=True, exist_ok=True)
    path = directory / f"nishikihebi-{timestamp:%Y%m%dT%H%M%SZ}.jsonl"

    logger = logging.getLogger("nishikihebi")
    logger.setLevel(logging.DEBUG)
    logger.handlers.clear()

    console_handler = logging.StreamHandler()
    console_handler.setLevel(logging.INFO)
    console_handler.setFormatter(ColorFormatter(console_handler.stream))
    logger.addHandler(console_handler)

    file_handler = logging.FileHandler(path)
    file_handler.setLevel(logging.DEBUG)
    file_handler.setFormatter(JsonLinesFormatter())
    logger.addHandler(file_handler)

    return path


class ContextLogger:
    def __init__(self, logger: logging.Logger) -> None:
        self.logger = logger

    def debug(self, message: str, **fields: object) -> None:
        self.logger.debug(message, extra={"context": fields}, stacklevel=2)

    def info(self, message: str, **fields: object) -> None:
        self.logger.info(message, extra={"context": fields}, stacklevel=2)

    def warning(self, message: str, **fields: object) -> None:
        self.logger.warning(message, extra={"context": fields}, stacklevel=2)

    def error(self, message: str, **fields: object) -> None:
        self.logger.error(message, extra={"context": fields}, stacklevel=2)


def get_logger(name: str) -> ContextLogger:
    return ContextLogger(logging.getLogger(name))

File: src/logs.ts
import { appendFileSync, mkdirSync } from "node:fs";
import { join } from "node:path";
import {
  BOLD,
  CYAN,
  colorEnabled,
  DIM,
  type OutputStream,
  RED,
  style,
  YELLOW,
} from "./console.js";

export type LogLevel = "DEBUG" | "INFO" | "WARNING" | "ERROR" | "CRITICAL";

export type LogContext = Readonly<Record<string, unknown>>;

export interface LogRecord {
  readonly time: string;
  readonly level: LogLevel;
  readonly logger: string;
  readonly message: string;
  readonly context: LogContext;
}

export interface LogHandler {
  handle(record: LogRecord): void;
}

const LEVEL_ORDER: Record<LogLevel, number> = {
  DEBUG: 10,
  INFO: 20,
  WARNING: 30,
  ERROR: 40,
  CRITICAL: 50,
};

const LEVEL_CODES: Record<LogLevel, readonly string[]> = {
  DEBUG: [DIM],
  INFO: [CYAN],
  WARNING: [YELLOW],
  ERROR: [RED],
  CRITICAL: [BOLD, RED],
};

/** Renders a record the way the console handler does: `LEVEL   message`. */
export function formatConsoleRecord(record: LogRecord, stream: OutputStream): string {
  const padded = record.level.padEnd(7);
  const levelName = colorEnabled(stream)
    ? style(padded, LEVEL_CODES[record.level], stream)
    : padded;
  return `${levelName} ${record.message}`;
}

class ConsoleHandler implements LogHandler {
  constructor(
    private readonly stream: OutputStream,
    private readonly minimumLevel: LogLevel,
  ) {}

  handle(record: LogRecord): void {
    if (LEVEL_ORDER[record.level] < LEVEL_ORDER[this.minimumLevel]) {
      return;
    }
    this.stream.write(`${formatConsoleRecord(record, this.stream)}\n`);
  }
}

class JsonLinesHandler implements LogHandler {
  constructor(private readonly path: string) {}

  handle(record: LogRecord): void {
    const payload = {
      time: record.time,
      level: record.level,
      logger: record.logger,
      message: record.message,
      ...record.context,
    };
    appendFileSync(this.path, `${JSON.stringify(payload)}\n`);
  }
}

let handlers: LogHandler[] = [];

/** Replaces the installed handlers. Test seam. */
export function setHandlers(next: readonly LogHandler[]): void {
  handlers = [...next];
}

/** Removes every handler, so records are dropped. Test seam. */
export function resetHandlers(): void {
  handlers = [];
}

function stamp(timestamp: Date): string {
  return `${timestamp.toISOString().slice(0, 19).replace(/[-:]/g, "")}Z`;
}

export interface LoggingOptions {
  readonly directory?: string;
  readonly timestamp?: Date;
  readonly stream?: OutputStream;
}

/**
 * Sends `DEBUG` and above to `<directory>/angel-<timestamp>.jsonl` and `INFO`
 * and above to the console, and returns the file path.
 */
export function configureLogging(options: LoggingOptions = {}): string {
  const directory = options.directory ?? "log";
  const timestamp = options.timestamp ?? new Date();
  const stream = options.stream ?? process.stderr;

  mkdirSync(directory, { recursive: true });
  const path = join(directory, `angel-${stamp(timestamp)}.jsonl`);

  setHandlers([new ConsoleHandler(stream, "INFO"), new JsonLinesHandler(path)]);
  return path;
}

/** A logger that takes structured fields alongside the message. */
export class ContextLogger {
  constructor(readonly name: string) {}

  private emit(level: LogLevel, message: string, context: LogContext): void {
    const record: LogRecord = {
      time: new Date().toISOString(),
      level,
      logger: this.name,
      message,
      context,
    };
    for (const handler of handlers) {
      handler.handle(record);
    }
  }

  debug(message: string, context: LogContext = {}): void {
    this.emit("DEBUG", message, context);
  }

  info(message: string, context: LogContext = {}): void {
    this.emit("INFO", message, context);
  }

  warning(message: string, context: LogContext = {}): void {
    this.emit("WARNING", message, context);
  }

  error(message: string, context: LogContext = {}): void {
    this.emit("ERROR", message, context);
  }
}

/** Returns the logger named for the calling module. */
export function getLogger(name: string): ContextLogger {
  return new ContextLogger(name);
}
